src/graph.py

```
@@ -0,0 +1,84 @@
+from typing import Any, Dict, List, Optional, Tuple, Union
+
+import numpy as np
+
+
+class Graph:
+    def __init__(
+        self, init_graph: Optional[Union[Dict[Any, List[Any]], np.ndarray]] = None
+    ) -> None:
+        """Initialize a graph, optionally from an adjacency list or matrix.
+
+        Args:
+            init_graph: Optional initialization graph, either as:
+                       - Dict[node, List[neighbors]] (adjacency list)
+                       - np.ndarray (adjacency matrix)
+        """
+        self.nodes: Dict[Any, List[Any]] = {}
+
+        if init_graph is not None:
+            if isinstance(init_graph, dict):
+                # Initialize from adjacency list
+                for node, neighbors in init_graph.items():
+                    self.add_node(node)
+                    for neighbor in neighbors:
+                        self.add_node(neighbor)
+                        self.add_edge(node, neighbor)
+            elif isinstance(init_graph, np.ndarray):
+                # Initialize from adjacency matrix
+                n = len(init_graph)
+                for i in range(n):
+                    self.add_node(i)
+                    for j in range(n):
+                        if init_graph[i][j]:
+                            self.add_edge(i, j)
+
+    def add_node(self, value: Any) -> None:
+        """Add a node to the graph.
+
+        Args:
+            value: The value of the node to be added.
+        """
+        if value not in self.nodes:
```

Better to raise an exception if a client tries to add already existed node rather then silently ingore it